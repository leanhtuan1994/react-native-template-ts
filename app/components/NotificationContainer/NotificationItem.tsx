import React from "react";
import { useWindowDimensions } from "react-native";
import { Text } from "react-native-elements";
import Animated, { Easing } from "react-native-reanimated";
import { Colors, FontSize, FontWeight } from "styles";
import { GlobalStyles, LayoutPadding } from "styles/theme";

export type Nullable<T> = T | null | undefined;

export type Notification = {
  title?: string;
  subtitle: string;
};

type Props = {
  notification?: Notification;
  onRemove(): void;
  offsetBottom?: number;
};

const easing = Easing.inOut(Easing.sin);

export const NotificationItem: React.FC<Props> = ({
  notification,
  onRemove,
  offsetBottom = 16,
}) => {
  const animationValue = React.useRef(new Animated.Value(0)).current;

  React.useEffect(() => {
    if (!notification) {
      return;
    }

    const showConfig = { toValue: 1, easing, duration: 250 };
    const delayConfig = { toValue: 1, easing, duration: 1000 };
    const hideConfig = { toValue: 0, easing, duration: 250 };

    let anim: Nullable<Animated.BackwardCompatibleWrapper> = Animated.timing(
      animationValue,
      showConfig,
    );

    anim.start((show) => {
      if (!show.finished) {
        return;
      }

      anim = Animated.timing(animationValue, delayConfig);
      anim.start((delay) => {
        if (!delay.finished) {
          return;
        }

        anim = Animated.timing(animationValue, hideConfig);
        anim.start((hide) => {
          if (!hide.finished) {
            return;
          }
          anim = null;
          onRemove();
        });
      });
    });

    return () => {
      if (anim) {
        anim.stop();
      }
      onRemove();
    };
  }, [animationValue, notification, onRemove]);

  const { width } = useWindowDimensions();

  const translateY = animationValue.interpolate({
    inputRange: [0, 1],
    outputRange: [30, 0],
  });

  const scale = animationValue.interpolate({
    inputRange: [0, 1],
    outputRange: [0.8, 1],
  });

  return (
    <Animated.View
      pointerEvents="none"
      style={{
        position: "absolute",
        bottom: offsetBottom,
        left: LayoutPadding.horizontal,
        width: width - LayoutPadding.horizontal * 2,
        padding: 12,
        borderRadius: 8,
        justifyContent: "center",
        backgroundColor: Colors.Black,
        transform: [
          {
            translateY,
          },
          {
            scale,
          },
        ],
        opacity: animationValue,
      }}>
      <Text
        style={GlobalStyles.font({
          fontSize: FontSize.Small,
          color: Colors.Secondary,
          fontWeight: FontWeight.SemiBold,
        })}>
        {notification ? notification.title : ""}
      </Text>
      <Text
        style={GlobalStyles.font({
          color: Colors.White,
          fontWeight: FontWeight.SemiBold,
        })}>
        {notification ? notification.subtitle : ""}
      </Text>
    </Animated.View>
  );
};
