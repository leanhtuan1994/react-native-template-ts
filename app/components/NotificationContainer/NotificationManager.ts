import { EventEmitter } from "events";

import { Notification } from "./NotificationItem";

const Constants = {
  CHANGE: "change",
};

class NotificationManager extends EventEmitter {
  create(notify: Notification) {
    this.emit(Constants.CHANGE, notify);
  }

  addChangeListener(callback: (notify: Notification | undefined) => void) {
    this.addListener(Constants.CHANGE, callback);
  }

  removeChangeListener(callback: (notify: Notification | undefined) => void) {
    this.removeListener(Constants.CHANGE, callback);
  }
}

const ins = new NotificationManager();

export default ins;
