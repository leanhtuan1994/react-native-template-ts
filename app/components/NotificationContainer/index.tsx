import React from "react";

import { Notification, NotificationItem } from "./NotificationItem";
import NotificationManager from "./NotificationManager";

type Props = {
  offsetBottom?: number;
};

const NotificationContainer: React.FC<Props> = ({ offsetBottom }) => {
  const [notification, setNotification] = React.useState<
    Notification | undefined
  >();

  const removeNotification = React.useCallback(() => {
    console.log("remove notification");
  }, []);

  React.useEffect(() => {
    NotificationManager.addChangeListener(setNotification);
    return () => {
      NotificationManager.removeChangeListener(setNotification);
    };
  }, []);

  return (
    <NotificationItem
      notification={notification}
      onRemove={removeNotification}
      offsetBottom={offsetBottom}
    />
  );
};

export default NotificationContainer;
