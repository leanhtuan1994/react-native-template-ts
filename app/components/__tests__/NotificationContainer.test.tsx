import React from "react";
import { shallow } from "enzyme";

import NotificationContainer from "../NotificationContainer";

describe("NotificationContainer", () => {
  it("notification match to snapshoot", () => {
    const notification = shallow(<NotificationContainer />);
    expect(notification).toMatchSnapshot();
  });
});
