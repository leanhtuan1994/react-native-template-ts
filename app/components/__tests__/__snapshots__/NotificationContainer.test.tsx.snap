// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationContainer notification match to snapshoot 1`] = `
<NotificationItem
  onRemove={[Function]}
/>
`;
